tests/test_serialization.py

```diff
@@ -627,3 +642,161 @@ class RootContainer:
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class ParentClass(Generic[_TData]):
+        value: str
+        data: _TData
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class ChildClass(ParentClass[Data[int]]):
+        pass
+
+    instance = ChildClass(value="vvv", data=Data(xxx=111))
+    dumped = mr.dump(instance)
+
+    assert dumped == {"value": "vvv", "data": {"xxx": 111}}
+    assert mr.load(ChildClass, dumped) == instance
+
+
+def test_generic_type_var_with_reuse() -> None:
+    _T = TypeVar("_T")
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class T1(Generic[_T]):
+        t1: _T
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class T2(Generic[_T], T1[int]):
+        t2: _T
+
+    instance = T2[str](t1=1, t2="2")
+
+    dumped = mr.dump(instance, cls=T2[str])
+
+    assert dumped == {"t1": 1, "t2": "2"}
+    assert mr.load(T2[str], dumped) == instance
+
+
+def test_generic_with_field_override() -> None:
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class Value1:
+        v1: str
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class Value2(Value1):
+        v2: str
+
+    _TValue = TypeVar("_TValue", bound=Value1)
+    _TItem = TypeVar("_TItem")
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class T1(Generic[_TItem]):
+        value: Value1
+        iterable: Iterable[_TItem]
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class T2(Generic[_TValue, _TItem], T1[_TItem]):
+        value: _TValue
+        iterable: set[_TItem]
+
+    instance = T2[Value2, int](value=Value2(v1="aaa", v2="bbb"), iterable=set([3, 4, 5]))
+
+    dumped = mr.dump(instance, cls=T2[Value2, int])
+
+    assert dumped == {"value": {"v1": "aaa", "v2": "bbb"}, "iterable": [3, 4, 5]}
+    assert mr.load(T2[Value2, int], dumped) == instance
+
+
+def test_generic_reuse_with_different_args() -> None:
+    _TItem = TypeVar("_TItem")
+
+    @dataclasses.dataclass(frozen=True, slots=True, kw_only=True)
+    class GenericContainer(Generic[_TItem]):
+        items: list[_TItem]
+
+    container_int = GenericContainer[int](items=[1, 2, 3])
+    dumped = mr.dump(container_int, cls=GenericContainer[int])
+
+    assert dumped == {"items": [1, 2, 3]}
+    assert mr.load(GenericContainer[int], dumped) == container_int
+
+    container_str = GenericContainer[str](items=["q", "w", "e"])
+    dumped = mr.dump(container_str, cls=GenericContainer[str])
+
+    assert dumped == {"items": ["q", "w", "e"]}
+    assert mr.load(GenericContainer[str], dumped) == container_str
+
+
+def test_sdfdfsd():
+    import dataclasses
+    from typing import Generic, TypeVar
+
+    import marshmallow_recipe as mr
+
+    T = TypeVar("T")
+
+    @dataclasses.dataclass()
+    class Regular(Generic[T]):
+        value: T
+
+    mr.dump(Regular[int](value=123))  # it works without explicit cls arg
```

Comment:
a call without type argument `mr.dump(Regular(value=123))` raises exception `ValueError(f"Expected subscripted generic, but got unsubscripted {t}")`, and it could be not so obvious from user perspective what went wrong. I would suggest to adjust error message with additional phrase like `...type arguments should be specified explicitly for generic type`

Reply:
improved https://github.com/anna-money/marshmallow-recipe/pull/163/commits/7a944cd1a8e251e672837b74b7141f47f6016d26